source/client/service_impl.cc

```diff
@@ -107,5 +109,65 @@ ::grpc::Status ServiceImpl::ExecutionStream(
   return finishGrpcStream(true);
 }
 
+namespace {
+void addHeader(envoy::api::v2::core::HeaderMap* map, absl::string_view key,
+               absl::string_view value) {
+  auto* request_header = map->add_headers();
+  request_header->set_key(std::string(key));
+  request_header->set_value(std::string(value));
+}
+} // namespace
+
+RequestSourcePtr RequestSourceServiceImpl::createStaticEmptyRequestSource(const uint32_t amount) {
+  Envoy::Http::HeaderMapPtr header = std::make_unique<Envoy::Http::HeaderMapImpl>();
+  header->addCopy(Envoy::Http::LowerCaseString("x-from-remote-request-source"), "1");
+  return std::make_unique<StaticRequestSourceImpl>(std::move(header), amount);
+}
+
+::grpc::Status RequestSourceServiceImpl::RequestStream(
+    ::grpc::ServerContext* /*context*/,
+    ::grpc::ServerReaderWriter<::nighthawk::request_source::RequestStreamResponse,
+                               ::nighthawk::request_source::RequestStreamRequest>* stream) {
+  nighthawk::request_source::RequestStreamRequest request;
+  bool ok = true;
+  while (stream->Read(&request)) {
+    ENVOY_LOG(trace, "Inbound RequestStreamRequest {}", request.DebugString());
+
+    // TODO(oschaaf): this is useful for integration testing purposes, but sending
```

Comment:
I finally understand what's going on here. So this is the implementation of a trivial request source gRPC service, and internally it instantiates yet another request source, but a static in-process one. And this request source server code ends up performing the same request_generator() calls that the nighthawk_client does.

I think we should assume that someone else could also be confused by this situation, even if they figure it out 10x faster. We might want to add a comment calling out the fact that the usage here is a clever one, and the one benchmark_client_impl.cc is a normal one.

And now that I think about it more, I don't see why future enhancements to Nighthawk like reading from a file should be placed in the RequestSource service, behind a layer of indirection through gRPC -- especially if the file reading will be implemented as a RequestSource anyway, and the RequestSource factory could just instantiate the file-reading RequestSource in the nighthawk_client process directly.

Reply:
yeah, exactly. So the current state should be taken as a first step, and when pulling traffic in over gRPC via the remote request source it is just a (very elaborate) no-op when comparing what nighthawk_client does today. 

But it has some use, in proving that it all works, landing test coverage, and also that the new abstractions we introduced while working on this are on par for the final destination we have in mind. 

When I wrote this, what I had in mind is that we would have a request source abstraction, and users can plug implementations of that into either nighthawk client or nighthawk service. The idea behind that is that we build generic tools and let users decide how and where use them. I think that when we get to factorizing `RequestSource` based on options, this will automatically propagate to both the `nighthawk_client` and `nighthawk_service`. Is that what you are suggesting as well?

Reply:
Yes. I think we're good with the code in this PR since it leaves it open to handle each of those potential RequestSource implementations in its own way.